website/blog/2022-10-18-how-to-design-and-structure-metrics.md

```diff
@@ -0,0 +1,381 @@
+
+The solution is to use the flexibility of the dbt modeling layer to add a boolean field called `is_active_row` that shows whether or not the row in question is the most recent version. With this, we can understand and filter out duplicate rows that may be connected to the same order.
+
+Once we have this field, we reach a diverging path:
+
+- If we are not interested in seeing the history of `order_events` , we can add a `where` clause **to the model itself**. This would ensure there is only one row per order.
+- If we **are** interested in seeing the history of `order_events` , we can add a `filter` to the metric definition to ensure that these duplicate order rows don’t cause us to misreport revenue
+
+Both of these paths ensure that only the correct orders are included in the metric calculation but one does it at the modeling layer and the other the semantic layer. There is no **best** path here - it depends on your organizations reporting and analytics needs. For this example, we’ll say that our business isn’t interested in understanding orders that have gone from completed to returned and so we’ll use option one moving forward. Now lets define the metric:
+
+```yaml
+version: 2
+metrics:
+  - name: total_revenue
+    label: The revenue of our business
+    model: ref('order_events')
+    description: "The revenue for our business, as defined by Jerry in Finance"
+
+    calculation_method: sum
+    expression: amount 
+
+    timestamp: order_date
+    time_grains: [day, week, month, all_time]
+
+    dimensions:
+      - customer_status
+			- order_country
+    
+    ## We don't need this section because we chose option 1
+    ## filters:
+    ##   - field: order_status
+    ##     operator: '='
+    ##     value: 'completed
+```
+
+Each of the properties of the above definition are defined [in the metrics documentation](https://docs.getdbt.com/docs/building-a-dbt-project/metrics), but let’s dig into the two that might require some additional explanation. The two in question are `expression` and `dimensions`.
+
+In plain english, the `expression` property is the sql column (or expression) that we are applying the calculation method on. In our example above, this simplifies to `sum(amount)`. However, this doesn’t **need** to be a field in the model. It could also be a sql expression like `case when condition = true then field else 0 end` .
+
+And then there’s `dimensions`.
+
+### Choosing which dimensions to use with your metric
+
+The `dimensions` attribute is a bit more nuanced than the others because it involves curating the ways through which a user can interact with the metric. To that end …
+
+*❗ **We recommend curating dimensions, not including all columns within the model. Most models contain dimensions that aren’t relevant for end-user analysis.***
```

Comment:
There's something wonky happening here with the asterisks again, looks like there's an extra one hanging out in the beginning?